src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/WitnessCommitmentsRule.cs

```diff
@@ -104,37 +112,151 @@ private bool EqualsArray(byte[] a, byte[] b, int length)
+            Script commitScriptPubKey = null;
+
+            if (network.Consensus.IsProofOfStake && network.Consensus.PosEmptyCoinbase)
             {
-                Script scriptPubKey = block.Transactions[0].Outputs[i].ScriptPubKey;
+                Script scriptPubKey = block.Transactions[0].Inputs[0].ScriptSig;
+
+                var ops = scriptPubKey.ToOps();
 
-                if (scriptPubKey.Length >= 38)
+                if (ops.Count > 2 && IsWitnessScript(new Script(ops.Skip(2))))
+                {
+                    // We assume the first two ops are the BIP34 coinbase height.
+                    commitScriptPubKey = new Script(ops.Skip(2));
+                }
+            }
+            else
+            {
+                for (int i = 0; i < block.Transactions[0].Outputs.Count; i++)
+                {
+                    Script scriptPubKey = block.Transactions[0].Outputs[i].ScriptPubKey;
+
+                    if (IsWitnessScript(scriptPubKey))
+                    {
+                        commitScriptPubKey = scriptPubKey;
+                    }
+                }
+            }
+
+            return commitScriptPubKey;
+        }
+
+        /// <summary>
+        /// Clear all witness commitments form the block.
+        /// </summary>
+        public static void ClearWitnessCommitment(Network network, Block block)
+        {
+            if (network.Consensus.IsProofOfStake && network.Consensus.PosEmptyCoinbase)
+            {
+                Script scriptPubKey = block.Transactions[0].Inputs[0].ScriptSig;
+
+                var ops = scriptPubKey.ToOps();
+
+                if (ops.Count > 2 && IsWitnessScript(new Script(ops.Skip(2))))
+                {
+                    block.Transactions[0].Inputs[0].ScriptSig = new Script(ops.Take(2));
+                }
+            }
+            else
+            {
+                for (int i = 0; i < block.Transactions[0].Outputs.Count; i++)
                 {
-                    byte[] scriptBytes = scriptPubKey.ToBytes(true);
-
-                    if ((scriptBytes[0] == (byte)OpcodeType.OP_RETURN) &&
-                        (scriptBytes[1] == 0x24) &&
-                        (scriptBytes[2] == 0xaa) &&
-                        (scriptBytes[3] == 0x21) &&
-                        (scriptBytes[4] == 0xa9) &&
-                        (scriptBytes[5] == 0xed))
+                    Script scriptPubKey = block.Transactions[0].Outputs[i].ScriptPubKey;
+
+                    if (IsWitnessScript(scriptPubKey))
                     {
-                        commitpos = i;
+                        block.Transactions[0].Outputs.RemoveAt(i);
+                        i--;
                     }
                 }
             }
+        }
+
+        /// <summary>
+        /// Create a witness commitmnet based from the given block.
+        /// </summary>
+        public static void CreateWitnessCommitment(Network network, Block block)
```

Comment:
We need tests for this method

Reply:
Done

Reply:
Perhaps yes but its defo tested on integration tests, so less priority right now IMO.